syncstorage/src/tokenserver/handlers.rs

```diff
@@ -108,28 +112,72 @@ fn get_token_plaintext(
+        // If there is a keys_changed_at on the request but not one on the user record, this is the
+        // first time the client reported it, so we assign the new value.
+        (Some(request_keys_changed_at), None) => Some(request_keys_changed_at),
+        // At this point, we've already validated that, if there is a keys_changed_at already
+        // stored on the user record, there must be one in the request. If that isn't the case,
+        // we've already returned an error.
+        (None, Some(user_keys_changed_at)) if user_keys_changed_at != 0 => unreachable!(),
+        // If there's no keys_changed_at in the request and the keys_changed_at on the user record
+        // is 0, keep the value as 0.
+        (None, Some(_user_keys_changed_at)) => Some(0),
+        // If there is no keys_changed_at on the user record or in the request, we want to leave
+        // the value unset.
+        (None, None) => None,
+    };
+
+    let generation = match req.auth_data.generation {
+        // If there's a generation in the request and it's greater than or equal to that stored on
+        // the user record, update to the value in the request.
+        Some(request_generation) if request_generation >= req.user.generation => request_generation,
+        // If there's a generation in the request and it's smaller than that stored on the user
+        // record, we've already returned an error.
+        Some(_request_generation) => unreachable!(),
+        None => match (req.auth_data.keys_changed_at, req.user.keys_changed_at) {
+            // If there's not a generation on the request but the keys_changed_at on the request
+            // is greater than the user's current generation AND the keys_changed_at on the request
+            // is greater than the user's current keys_changed_at, set the user's generation to
+            // the new keys_changed_at.
+            (Some(request_keys_changed_at), Some(user_keys_changed_at))
+                if request_keys_changed_at > user_keys_changed_at
+                    && request_keys_changed_at > req.user.generation =>
+            {
+                request_keys_changed_at
+            }
+            // If there's not a generation on the request but the keys_changed_at on the request
+            // is greater than the user's current generation AND there is a keys_changed_at on the
+            // request but not currently on the user record, set the user's generation to the new
+            // keys_changed_at.
+            (Some(request_keys_changed_at), None)
+                if request_keys_changed_at > req.user.generation =>
+            {
+                request_keys_changed_at
+            }
+            // If the request has a keys_changed_at but the above conditions don't hold OR if the
+            // request doesn't have a keys_changed_at, just keep the same generation.
+            (_, _) => req.user.generation,
```

Comment:
Wouldn't this be the equivalent?

```suggestion
        None => match keys_changed_at {
            // No generation on the request but an updated keys_changed_at greater than the user's
            // current generation: set the user's generation to the new keys_changed_at.
            Some(keys_changed_at) if keys_changed_at > req.user.generation => keys_changed_at,
            // Otherwise just keep the same generation.
            _ => req.user.generation,
```

Granted I get you're after the verbosity here but it it's basically the same as the above check (minus the 0 case which wouldn't trigger the `> generation` anyway)

Reply:
FWIW, I'm ok with the verbosity, since it's outlining and documenting potential states and results (a bit like having a matrix). 
I suppose those could be captured in just the documentation, though.

Reply:
Yeah I considered simplifying down to that, but I think it hides a number of distinct cases. I thought it would be preferable to be as explicit as possible so each edge case is clear